### PR TITLE
Minor cleanups to BytesReferenceStreamInput

### PR DESCRIPTION
**Pull request by DaveCTurner**
Followup to #61681:

- reuse the current iterator in `reset()` if possible
- simply some integer-overflow-avoidance in `skip()`
- clarify some comments
- address some IntelliJ warnings

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/bytes/AbstractBytesReference.java [hunk 1]
```diff
@@ -293,8 +295,7 @@ protected void ensureCanReadBytes(int bytesToRead) throws EOFException {
 
         @Override
         public long skip(long n) throws IOException {
-            final int skip = (int) Math.min(Integer.MAX_VALUE, n);
-            final int numBytesSkipped =  Math.min(skip, length() - offset());
+            final int numBytesSkipped = ByteSizeUnit.BYTES.toIntBytes(Math.min(n, length() - offset()));
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/common/bytes/AbstractBytesReference.java:298]: Not sure I'm the biggest fan of this, it's just a slower method on the hot path (much larger and won't always inline perfectly I think) isn't it + it requires the reader to look into what `ByteSizeUnit.BYTES.toIntBytes` does when the previous version was crystal clear here?

**Review by original-brownbear — COMMENTED** [review]
Not sure about one point. Nice find with the iterator though :)

File: server/src/main/java/org/elasticsearch/common/bytes/AbstractBytesReference.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/common/bytes/AbstractBytesReference.java]: Yeah ok we don't really need this assertion. There's a latent bug in the original code if called with `n < Integer.MIN_VALUE`: the cast would overflow and maybe we'd skip too much, and also the callers are not all super-strict at checking we skipped what we wanted to. Changed to a different assertion and lighter code in 5bb4280.

**Review by original-brownbear — APPROVED** [review]
LGTM thanks David!

**Pull request merged**